README.md
# ADDSAuditTasks Module
## Get-ADDSActiveAccountAudit
### Synopsis
Active Directory Audit with Keyvault retrieval option.
### Syntax
```powershell

Get-ADDSActiveAccountAudit [-LocalDisk] [-AttachmentFolderPath <String>] [-ADDSAccountIsNotEnabled] [-DaysInactive <Int32>] [<CommonParameters>]

Get-ADDSActiveAccountAudit [-SendMailMessage] [-SMTPServer <String>] [-AttachmentFolderPath <String>] [-ADDSAccountIsNotEnabled] [-DaysInactive <Int32>] -UserName <String> [-Password <SecureString>] [-Port <Int32>] -To <String> [-From <String>] [<CommonParameters>]

Get-ADDSActiveAccountAudit [-WinSCP] [-AttachmentFolderPath <String>] [-ADDSAccountIsNotEnabled] [-DaysInactive <Int32>] -UserName <String> -Password <SecureString> -FTPHost <String> -SshHostKeyFingerprint <String> [-RemotePath <String>] [<CommonParameters>]

Get-ADDSActiveAccountAudit [-FunctionApp] <String> [-Function] <String> [-SMTPServer <String>] [-AttachmentFolderPath <String>] [-ADDSAccountIsNotEnabled] [-DaysInactive <Int32>] -UserName <String> [-Port <Int32>] -To <String> [-From <String>] -ApiToken <String> [<CommonParameters>]

Get-ADDSActiveAccountAudit [-Clean] [<CommonParameters>]




```
### Parameters
| Name  | Alias  | Description | Required? | Pipeline Input | Default Value |
| - | - | - | - | - | - |
| <nobr>LocalDisk</nobr> |  | Only output data to local disk. | true | false | False |
| <nobr>SendMailMessage</nobr> |  | Adds parameters for sending Audit Report as an Email. | true | false | False |
| <nobr>WinSCP</nobr> |  | Adds parameters for sending Audit Report via SFTP. | true | true \(ByPropertyName\) | False |
| <nobr>FunctionApp</nobr> |  | Azure Function App Name. | true | false |  |
| <nobr>Function</nobr> |  | Azure Function App's Function Name. Ex. "HttpTrigger1" | true | true \(ByPropertyName\) |  |
| <nobr>SMTPServer</nobr> |  | Defaults to Office 365 SMTP relay. Enter optional relay here. | false | true \(ByPropertyName\) | smtp.office365.com |
| <nobr>AttachmentFolderPath</nobr> |  | Default path is C:\\temp\\ADDSActiveAccountAuditLogs. This is the folder where attachments are going to be saved. | false | true \(ByValue\) | C:\\temp\\ADDSActiveAccountAuditLogs |
| <nobr>ADDSAccountIsNotEnabled</nobr> |  | Defaults to not being set. Choose to search for disabled Active Directory Users. | false | true \(ByPropertyName\) | False |
| <nobr>DaysInactive</nobr> |  | Defaults to 90 days in the past. Specifies how far back to look for accounts last logon. If logon is within 90 days, it won't be included. | false | true \(ByPropertyName\) | 90 |
| <nobr>UserName</nobr> |  | Specify the account with an active mailbox and MFA disabled. Ensure the account has delegated access for Send On Behalf for any UPN set in the "$From" Parameter | true | true \(ByPropertyName\) |  |
| <nobr>Password</nobr> |  | Use: \(Read-Host -AsSecureString\) as in Examples. May be omitted. | false | true \(ByPropertyName\) |  |
| <nobr>Port</nobr> |  | SMTP Port to Relay. Ports can be: "993", "995", "587", or "25" | false | true \(ByPropertyName\) | 587 |
| <nobr>To</nobr> |  | Recipient of the attachment outputs. | true | true \(ByPropertyName\) |  |
| <nobr>From</nobr> |  | Defaults to the same account as $UserName unless the parameter is set. Ensure the Account has delegated access to send on behalf for the $From account. | false | true \(ByPropertyName\) | $UserName |
| <nobr>ApiToken</nobr> |  | Private Function Key. | true | true \(ByPropertyName\) |  |
| <nobr>FTPHost</nobr> |  | SFTP Hostname. | true | true \(ByPropertyName\) |  |
| <nobr>SshHostKeyFingerprint</nobr> |  | Adds parameters for sending Audit Report via SFTP. | true | true \(ByPropertyName\) |  |
| <nobr>RemotePath</nobr> |  | Remove FTP path. Will be created in the user path under functionname folder if not specified. | false | true \(ByPropertyName\) | \("./" \\+ $\($MyInvocation.MyCommand.Name -replace '\\..\*'\)\) |
| <nobr>Clean</nobr> |  | Remove installed modules during run. Remove local files if not a LocalDisk run. | true | false | False |
### Note
Can take password as input into secure string using \(Read-Host -AsSecureString\).

### Examples
**EXAMPLE 1**
```powershell
Get-ADDSActiveAccountAudit -LocalDisk -Verbose
```


**EXAMPLE 2**
```powershell
Get-ADDSActiveAccountAudit -SendMailMessage -SMTPServer $SMTPServer -UserName "helpdesk@domain.com" -Password (Read-Host -AsSecureString) -To "support@domain.com" -Verbose
```


**EXAMPLE 3**
```powershell
Get-ADDSActiveAccountAudit -FunctionApp $FunctionApp -Function $Function -SMTPServer $SMTPServer -UserName "helpdesk@domain.com" -To "support@domain.com" -Verbose
```


**EXAMPLE 4**
```powershell
Get-ADDSActiveAccountAudit -WinSCP -UserName "ftphostname.UserName" -Password (Read-Host -AsSecureString) -FTPHost "ftphost.domain.com" -SshHostKeyFingerprint "<SShHostKeyFingerprint>" -Verbose
```


**EXAMPLE 5**
```powershell
Get-ADDSActiveAccountAudit -Clean -Verbose
```


## Get-ADDSAssetInventoryAudit
### Synopsis
Active Directory Server and Workstation Audit with Report export option \(Can also be piped to CSV if Report isn't specified\).
### Syntax
```powershell

Get-ADDSAssetInventoryAudit [-HostType] <String> [[-DaystoConsiderAHostInactive] <Int32>] [[-Report]] [[-DirPath] <String>] [-Enabled <Boolean>] [<CommonParameters>]

Get-ADDSAssetInventoryAudit [-OSType] <String> [[-DaystoConsiderAHostInactive] <Int32>] [[-Report]] [[-DirPath] <String>] [-Enabled <Boolean>] [<CommonParameters>]




```
### Parameters
| Name  | Alias  | Description | Required? | Pipeline Input | Default Value |
| - | - | - | - | - | - |
| <nobr>HostType</nobr> |  | Select from Windows Server or Windows 10 plus. | true | true \(ByValue\) |  |
| <nobr>OSType</nobr> |  | Search an OS String. Wildcards can be omitted as the function will automatically add the wildcard characters before searching. | true | true \(ByValue\) |  |
| <nobr>DaystoConsiderAHostInactive</nobr> |  |  | false | true \(ByPropertyName\) | 90 |
| <nobr>Report</nobr> |  | Add report output as csv to DirPath directory. | false | true \(ByPropertyName\) | False |
| <nobr>DirPath</nobr> |  | The path to the -Report output directory. | false | false | C:\\temp\\ADDSAssetInventoryAudit |
| <nobr>Enabled</nobr> |  |  | false | true \(ByPropertyName\) | True |
### Note
Outputs to C:\\temp by default. For help type: help Get-ADDSAssetInventoryAudit -ShowWindow

### Examples
**EXAMPLE 1**
```powershell
Get-ADDSInventoryAudit -HostType WindowsServers
```


**EXAMPLE 2**
```powershell
Get-ADDSInventoryAudit -HostType WindowsWorkstations -DirPath "C:\Temp\" -Report
```


**EXAMPLE 3**
```powershell
Get-ADDSInventoryAudit -HostType WindowsServers -DirPath "C:\Temp\" -Report
```


**EXAMPLE 4**
```powershell
Get-ADDSInventoryAudit -OSType "2008" -DirPath "C:\Temp\" -Report
```


## Get-ADDSDepartedUsersAccountAudit
### Synopsis
Active Directory Audit with Keyvault retrieval option.
### Syntax
```powershell

Get-ADDSDepartedUsersAccountAudit [-LocalDisk] [-AttachmentFolderPath <String>] -WildCardIdentifier <String> [<CommonParameters>]

Get-ADDSDepartedUsersAccountAudit [-SendMailMessage] [-SMTPServer <String>] [-AttachmentFolderPath <String>] -UserName <String> [-Password <SecureString>] [-Port <Int32>] -To <String> [-From <String>] -WildCardIdentifier <String> [<CommonParameters>]

Get-ADDSDepartedUsersAccountAudit [-WinSCP] [-AttachmentFolderPath <String>] -UserName <String> -Password <SecureString> -FTPHost <String> -SshHostKeyFingerprint <String> [-RemotePath <String>] -WildCardIdentifier <String> [<CommonParameters>]

Get-ADDSDepartedUsersAccountAudit [-FunctionApp] <String> [-Function] <String> [-SMTPServer <String>] [-AttachmentFolderPath <String>] -UserName <String> [-Port <Int32>] -To <String> [-From <String>] -ApiToken <String> -WildCardIdentifier <String> [<CommonParameters>]

Get-ADDSDepartedUsersAccountAudit [-Clean] [<CommonParameters>]




```
### Parameters
| Name  | Alias  | Description | Required? | Pipeline Input | Default Value |
| - | - | - | - | - | - |
| <nobr>LocalDisk</nobr> |  | Only output data to local disk. | true | false | False |
| <nobr>SendMailMessage</nobr> |  | Adds parameters for sending Audit Report as an Email. | true | false | False |
| <nobr>WinSCP</nobr> |  | Adds parameters for sending Audit Report via SFTP. | true | true \(ByPropertyName\) | False |
| <nobr>FunctionApp</nobr> |  | Azure Function App Name. | true | false |  |
| <nobr>Function</nobr> |  | Azure Function App's Function Name. Ex. "HttpTrigger1" | true | true \(ByPropertyName\) |  |
| <nobr>SMTPServer</nobr> |  | Defaults to Office 365 SMTP relay. Enter optional relay here. | false | true \(ByPropertyName\) | smtp.office365.com |
| <nobr>AttachmentFolderPath</nobr> |  | Default path is C:\\temp\\ADDSDepartedUsersAuditLogs. This is the folder where attachments are going to be saved. | false | true \(ByValue\) | C:\\temp\\ADDSDepartedUsersAuditLogs |
| <nobr>UserName</nobr> |  | Specify the account with an active mailbox and MFA disabled. Ensure the account has delegated access for Send On Behalf for any UPN set in the "$From" Parameter | true | true \(ByPropertyName\) |  |
| <nobr>Password</nobr> |  | Use: \(Read-Host -AsSecureString\) as in Examples. May be omitted. | false | true \(ByPropertyName\) |  |
| <nobr>Port</nobr> |  | SMTP Port to Relay. Ports can be: "993", "995", "587", or "25" | false | true \(ByPropertyName\) | 587 |
| <nobr>To</nobr> |  | Recipient of the attachment outputs. | true | true \(ByPropertyName\) |  |
| <nobr>From</nobr> |  | Defaults to the same account as $UserName unless the parameter is set. Ensure the Account has delegated access to send on behalf for the $From account. | false | true \(ByPropertyName\) | $UserName |
| <nobr>ApiToken</nobr> |  | Private Function Key. | true | true \(ByPropertyName\) |  |
| <nobr>FTPHost</nobr> |  |  | true | true \(ByPropertyName\) |  |
| <nobr>SshHostKeyFingerprint</nobr> |  |  | true | true \(ByPropertyName\) |  |
| <nobr>RemotePath</nobr> |  |  | false | true \(ByPropertyName\) | \("./" \\+ $\($MyInvocation.MyCommand.Name -replace '\\..\*'\)\) |
| <nobr>Clean</nobr> |  | Remove installed modules during run. Remove local files if not a LocalDisk run. | true | false | False |
| <nobr>WildCardIdentifier</nobr> |  | Name wildcard appended to user account. | true | true \(ByPropertyName\) |  |
### Note
Can take password as input into secure string using \(Read-Host -AsSecureString\).

### Examples
**EXAMPLE 1**
```powershell
Get-ADDSDepartedUsersAccountAudit -LocalDisk -WildCardIdentifier "<StringToSearchFor>" -Verbose
```


**EXAMPLE 2**
```powershell
Get-ADDSDepartedUsersAccountAudit -SendMailMessage -SMTPServer $SMTPServer -UserName "helpdesk@domain.com" -Password (Read-Host -AsSecureString) -To "support@domain.com" -WildCardIdentifier "<StringToSearchFor>" -Verbose
```


**EXAMPLE 3**
```powershell
Get-ADDSDepartedUsersAccountAudit -FunctionApp $FunctionApp -Function $Function -SMTPServer $SMTPServer -UserName "helpdesk@domain.com" -To "support@domain.com" -WildCardIdentifier "<StringToSearchFor>" -Verbose
```


**EXAMPLE 4**
```powershell
Get-ADDSDepartedUsersAccountAudit -WinSCP -UserName "ftphostname.UserName" -Password (Read-Host -AsSecureString) -FTPHost "ftphost.domain.com" -SshHostKeyFingerprint "<SShHostKeyFingerprint>" -WildCardIdentifier "<StringToSearchFor>" -Verbose
```


**EXAMPLE 5**
```powershell
Get-ADDSDepartedUsersAccountAudit -Clean -Verbose
```


## Get-ADDSPrivilegedAccountAudit
### Synopsis
Active Directory Audit with Keyvault retrieval option.
### Syntax
```powershell

Get-ADDSPrivilegedAccountAudit [-LocalDisk] [-AttachmentFolderPath <String>] [<CommonParameters>]

Get-ADDSPrivilegedAccountAudit [-SendMailMessage] [-SMTPServer <String>] [-AttachmentFolderPath <String>] -UserName <String> [-Password <SecureString>] [-Port <Int32>] -To <String> [-From <String>] [<CommonParameters>]

Get-ADDSPrivilegedAccountAudit [-WinSCP] [-AttachmentFolderPath <String>] -UserName <String> -Password <SecureString> -FTPHost <String> -SshHostKeyFingerprint <String> [-RemotePath <String>] [<CommonParameters>]

Get-ADDSPrivilegedAccountAudit [-FunctionApp] <String> [-Function] <String> [-SMTPServer <String>] [-AttachmentFolderPath <String>] -UserName <String> [-Port <Int32>] -To <String> [-From <String>] -ApiToken <String> [<CommonParameters>]

Get-ADDSPrivilegedAccountAudit [-Clean] [<CommonParameters>]




```
### Parameters
| Name  | Alias  | Description | Required? | Pipeline Input | Default Value |
| - | - | - | - | - | - |
| <nobr>LocalDisk</nobr> |  | Only output data to local disk. | true | false | False |
| <nobr>SendMailMessage</nobr> |  | Adds parameters for sending Audit Report as an Email. | true | false | False |
| <nobr>WinSCP</nobr> |  | Adds parameters for sending Audit Report via SFTP. | true | true \(ByPropertyName\) | False |
| <nobr>FunctionApp</nobr> |  | Azure Function App Name. | true | false |  |
| <nobr>Function</nobr> |  | Azure Function App's Function Name. Ex. "HttpTrigger1" | true | true \(ByPropertyName\) |  |
| <nobr>SMTPServer</nobr> |  | Defaults to Office 365 SMTP relay. Enter optional relay here. | false | true \(ByPropertyName\) | smtp.office365.com |
| <nobr>AttachmentFolderPath</nobr> |  | Default path is C:\\temp\\ADDSPrivilegedAccountAuditLogs. This is the folder where attachments are going to be saved. | false | true \(ByValue\) | C:\\temp\\ADDSPrivilegedAccountAuditLogs |
| <nobr>UserName</nobr> |  | Specify the account with an active mailbox and MFA disabled. Ensure the account has delegated access for Send On Behalf for any UPN set in the "$From" Parameter | true | true \(ByPropertyName\) |  |
| <nobr>Password</nobr> |  | Use: \(Read-Host -AsSecureString\) as in Examples. May be omitted. | false | true \(ByPropertyName\) |  |
| <nobr>Port</nobr> |  | SMTP Port to Relay. Ports can be: "993", "995", "587", or "25" | false | true \(ByPropertyName\) | 587 |
| <nobr>To</nobr> |  | Recipient of the attachment outputs. | true | true \(ByPropertyName\) |  |
| <nobr>From</nobr> |  | Defaults to the same account as $UserName unless the parameter is set. Ensure the Account has delegated access to send on behalf for the $From account. | false | true \(ByPropertyName\) | $UserName |
| <nobr>ApiToken</nobr> |  | Private Function Key. | true | true \(ByPropertyName\) |  |
| <nobr>FTPHost</nobr> |  | SFTP Hostname. | true | true \(ByPropertyName\) |  |
| <nobr>SshHostKeyFingerprint</nobr> |  | Adds parameters for sending Audit Report via SFTP. | true | true \(ByPropertyName\) |  |
| <nobr>RemotePath</nobr> |  | Remove FTP path. Will be created in the user path under functionname folder if not specified. | false | true \(ByPropertyName\) | \("./" \\+ $\($MyInvocation.MyCommand.Name -replace '\\..\*'\)\) |
| <nobr>Clean</nobr> |  | Remove installed modules during run. Remove local files if not a LocalDisk run. | true | false | False |
### Note
Can take password as input into secure string using \(Read-Host -AsSecureString\).

### Examples
**EXAMPLE 1**
```powershell
Get-ADDSPrivilegedAccountAudit -LocalDisk -Verbose
```


**EXAMPLE 2**
```powershell
Get-ADDSPrivilegedAccountAudit -SendMailMessage -SMTPServer $SMTPServer -UserName "helpdesk@domain.com" -Password (Read-Host -AsSecureString) -To "support@domain.com" -Verbose
```


**EXAMPLE 3**
```powershell
Get-ADDSPrivilegedAccountAudit -FunctionApp $FunctionApp -Function $Function -SMTPServer $SMTPServer -UserName "helpdesk@domain.com" -To "support@domain.com" -Verbose
```


**EXAMPLE 4**
```powershell
Get-ADDSPrivilegedAccountAudit -WinSCP -UserName "ftphostname.UserName" -Password (Read-Host -AsSecureString) -FTPHost "ftphost.domain.com" -SshHostKeyFingerprint "<SShHostKeyFingerprint>" -Verbose
```


**EXAMPLE 5**
```powershell
Get-ADDSPrivilegedAccountAudit -Clean -Verbose
```


## Get-ADUsersLastLogon
### Synopsis
Takes SamAccountName as input to retrieve most recent LastLogon from all DC's.
### Syntax
```powershell

Get-ADUsersLastLogon [-SamAccountName] <Object> [<CommonParameters>]




```
### Parameters
| Name  | Alias  | Description | Required? | Pipeline Input | Default Value |
| - | - | - | - | - | - |
| <nobr>SamAccountName</nobr> | Identity, UserName, Account | The SamAccountName of the user being checked for LastLogon. | true | true \(ByValue\) |  |
### Outputs
 - System.DateTime

### Examples
**EXAMPLE 1**
```powershell
Get-ADUsersLastLogon -SamAccountName "UserName"
```


## Get-NetworkScan
### Synopsis
Discovers local network and runs port scans on all hosts found for specific or default sets of ports.
### Syntax
```powershell

Get-NetworkScan [[-Ports] <Int32[]>] [-LocalSubnets] [-Report] [<CommonParameters>]

Get-NetworkScan [[-Ports] <Int32[]>] [-Computers] <String[]> [-Report] [<CommonParameters>]




```
### Parameters
| Name  | Alias  | Description | Required? | Pipeline Input | Default Value |
| - | - | - | - | - | - |
| <nobr>Ports</nobr> |  | Default ports are: "21", "22", "23", "25", "53", "67", "68", "80", "443", \` "88", "464", "123", "135", "137", "138", "139", \` "445", "389", "636", "514", "587", "1701", \` "3268", "3269", "3389", "5985", "5986"  If you want to supply a port, do so as an integer or an array of integers. "22","80","443", etc. | false | true \(ByPropertyName\) |  |
| <nobr>LocalSubnets</nobr> |  |  | true | true \(ByPropertyName\) | False |
| <nobr>Computers</nobr> |  |  | true | true \(ByPropertyName\) |  |
| <nobr>Report</nobr> |  | Specify this switch if you would like a report generated in C:\\temp. | false | false | False |
### Note
Installs PSnmap if not found and can output a report, or just the results.

### Examples
**EXAMPLE 1**
```powershell
Get-NetworkScan -report
```


### Links

 - [Specify a URI to a help page, this will show when Get-Help -Online is used.](#Specify a URI to a help page, this will show when Get-Help -Online is used.)
## Switch-SurnameWithGivenName
### Synopsis
Takes CSV input as "LastName\<space\>FirstName" and flips it to "Firstname\<space\>Lastname"
### Syntax
```powershell

Switch-SurnameWithGivenName [[-RosterCSV] <String>] [[-AttachmentFolder] <String>] [<CommonParameters>]




```
### Parameters
| Name  | Alias  | Description | Required? | Pipeline Input | Default Value |
| - | - | - | - | - | - |
| <nobr>RosterCSV</nobr> |  |  | false | true \(ByValue, ByPropertyName\) |  |
| <nobr>AttachmentFolder</nobr> |  |  | false | true \(ByPropertyName\) | C:\\temp\\Switch-SurnameWithGivenName |
### Note
This function depends on the name column in the employee roster name column, to have been formatted in excel using a find and replace to replace ", " with " ". In other words: The file needs to have "comma space" replaces with "space" in the name column to be easily compared to ADUser output.

### Examples
**EXAMPLE 1**
```powershell
Switch-SurnameWithGivenName -RosterCSV "C:\temp\RosterNameColumnFormattedLastNameSpaceFirstname.csv" -Verbose
```


### Links

 - [Specify a URI to a help page, this will show when Get-Help -Online is used.](#Specify a URI to a help page, this will show when Get-Help -Online is used.)


CHANGELOG.md
# Changelog for ADDSAuditTasks

The format is based on and uses the types of changes according to [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Fixed

- Asset Inventory Ouput showing SPNs.

## [1.9.11] - 2022-12-15

### Added

- Active Directory Module dependancy

## [1.9.9] - 2022-12-13

### Added

- More generic call for workstations in asset inventory.
- Additional call for Non-Windows Machines using POSIX.

## [1.9.8] - 2022-12-11

### Added

- Reporting QoL improvements to Asset inventory
- Added zip and directory to Report Output
- Option to only provide output to console.
- Check for Active Directory Module
- Option to search enabled or disabled hosts in `Get-ADDSAssetInventoryAudit`
- Verbose Output to assist with identifying output location.

### Fixed

- Fixed `Get-ADDSActiveAccountAudit` so that output is piped to `$null`

## [1.9.7] - 2022-12-09

### Fixed

- Comment-Based help

## [1.9.6] - 2022-12-08

### Added

- Feature Added: Function `ADDSAssetInventoryAudit` with reporting.

## [1.9.5] - 2022-11-29

### Added

- Added telnet to default ports to scan.
- Added OUI lookup.
- Added MACID lookup

### Fixed

- Fixed `[ADAuditUser]` Class Department property.

## [1.9.4] - 2022-11-17

### Fixed

- Formatting in various modules. No function changes.

### Added

- Added options to scan specific hosts, subnets, IPs or FQDNs.

## [1.9.2] - 2022-11-08

### Fixed

- Fixed ordered hashtable in departed users audit.
- Fixed service principal name output.

## [1.9.0] - 2022-11-08

### Fixed

- Fixed line 36 of `Switch-SurnameWithGivenName` to `$HRCSV = Import-Csv $RosterCSV`

## [1.8.0] - 2022-11-08

### Removed

- Removed function `Format-HRRoster` to public functions to format roster for compare.

### Added

- Renamed function `Format-HRRoster` for clarity to `Switch-SurnameWithGivenName`.
- Added `Get-NetworkScan` public function to scan subnets for hosts and open ports.

## [1.6.1] - 2022-07-21

### Added

- Added Name parameter to `Get-ADDSActiveUsersAudit` and `Get-ADDSDepartedUsersAudit`

## [1.5.4] - 2022-07-21

### Fixed

- Fixed Log output of ADDSActiveUserAudit function.
- Fixed Release

## [1.5.0] - 2022-05-16

### Fixed

- Fixed Timestamp unassigned variable.
- Fixed unassigned time variables in public functions.

## [1.4.0] - 2022-05-15

### Fixed

- Fixed output for export in Privileged Account Audit.

## [1.3.0] - 2022-05-15

### Added

- Added FTP submission error handling.
- Added Class `[ADAuditAccount]`

## [1.2.0] - 2022-05-14

### Added

- Added Upload via WinSCP using private function `Submit-FTPUpload`

## [1.1.0] - 2022-05-11

### Added

- Added proper attachment handling with multiple files.

## [1.1.0-preview0001] - 2022-05-11

### Added

- Added proper attachment handling with multiple files.

## [0.7.0] - 2022-05-11

### Added

- Added fixes to zip handling.

## [0.6.5] - 2022-05-11

### Added

- Added extended rights to privileged account audit.
- Added private function `Get-ADExtendedRight`

### Changed

- Changed how extended user rights are obtained.
- Changed CSV and ZIP files names to sort by date.

### Removed

- Removed old method of retrieving AD privileged audit output.

### Fixed

- Fixed `Get-ADDSDepartedUsersAccountAudit` Not requiring wildcard.
- Log output for csv export.
- Log output filename for log file.

### Security

- In case of vulnerabilities.

## [0.6.4] - 2022-05-09

### Fixed

- Fixed unused parameters in public functions.


source/Public/Get-NetworkScan.ps1
function Get-NetworkScan {
    <#
    .SYNOPSIS
        Discovers local network and runs port scans on all hosts found for specific or default sets of ports.
    .DESCRIPTION
        Scans the network for open ports specified by the user or default ports if no ports are specified.
        Creates reports if report switch is active.
    .NOTES
        Installs PSnmap if not found and can output a report, or just the results.
    .LINK
        Specify a URI to a help page, this will show when Get-Help -Online is used.
    .EXAMPLE
        Get-NetworkScan -report
    .PARAMETER Ports
        Default ports are:
        "21", "22", "23", "25", "53", "67", "68", "80", "443", `
        "88", "464", "123", "135", "137", "138", "139", `
        "445", "389", "636", "514", "587", "1701", `
        "3268", "3269", "3389", "5985", "5986"

        If you want to supply a port, do so as an integer or an array of integers.
        "22","80","443", etc.
    .PARAMETER Report
        Specify this switch if you would like a report generated in C:\temp.
    #>
    [CmdletBinding(DefaultParameterSetName = 'Default' , HelpURI = "https://criticalsolutionsnetwork.github.io/ADDSAuditTasks/#Get-NetworkScan")]
    param (
        [Parameter(
            ValueFromPipelineByPropertyName = $true,
            Position = 0
        )]
        [ValidateRange(1, 65535)]
        [int[]]$Ports,
        [Parameter(
            Mandatory = $true,
            ParameterSetName = 'Default',
            HelpMessage = 'Automatically find and scan local attached subnets',
            ValueFromPipelineByPropertyName = $true,
            Position = 1
        )]
        [switch]$LocalSubnets,
        [Parameter(
            Mandatory = $true,
            ParameterSetName = 'Computers',
            HelpMessage = 'Scan host or array of hosts using Subet ID in CIDR Notation, IP, NETBIOS, or FQDN in "quotes"',
            ValueFromPipelineByPropertyName = $true,
            Position = 1
        )]
        [string[]]$Computers,
        [switch]$Report
    )
    begin {
        If (Get-Module -ListAvailable -Name "PSnmap") { Import-Module "PSnmap" } Else { Install-Module "PSnmap" -Force; Import-Module "PSnmap" }
        if (!($ports)) {
            [int[]]$ports = "21", "22", "23", "25", "53", "67", "68", "80", "443", `
                "88", "464", "123", "135", "137", "138", "139", `
                "445", "389", "636", "514", "587", "1701", `
                "3268", "3269", "3389", "5985", "5986"
        }
        $ouiobject = Invoke-RestMethod https://standards-oui.ieee.org/oui/oui.csv | ConvertFrom-Csv
    } # Begin Close
    process {
        if ($LocalSubnets) {
            $ConnectedNetworks = Get-NetIPConfiguration -Detailed | Where-Object { $_.Netadapter.status -eq "up" }
            $results = @()
            foreach ($network in $ConnectedNetworks) {
                # Get Network DHCP Server
                $DHCPServer = (Get-CimInstance -ClassName Win32_NetworkAdapterConfiguration | Where-Object { $_.IPAddress -eq $network.IPv4Address }).DHCPServer
                # Get Subnet as CIDR
                $Subnet = "$($network.IPv4DefaultGateway.nexthop)/$($network.IPv4Address.PrefixLength)"
                # Regex for IPV4 and IPV6 validation
                if (($subnet -match '^([0-9]{1,3}\.){3}[0-9]{1,3}(\/([0-9]|[1-2][0-9]|3[0-2]))?$') -or ($subnet -match '^s*((([0-9A-Fa-f]{1,4}:){7}([0-9A-Fa-f]{1,4}|:))|(([0-9A-Fa-f]{1,4}:){6}(:[0-9A-Fa-f]{1,4}|((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3})|:))|(([0-9A-Fa-f]{1,4}:){5}(((:[0-9A-Fa-f]{1,4}){1,2})|:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3})|:))|(([0-9A-Fa-f]{1,4}:){4}(((:[0-9A-Fa-f]{1,4}){1,3})|((:[0-9A-Fa-f]{1,4})?:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){3}(((:[0-9A-Fa-f]{1,4}){1,4})|((:[0-9A-Fa-f]{1,4}){0,2}:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){2}(((:[0-9A-Fa-f]{1,4}){1,5})|((:[0-9A-Fa-f]{1,4}){0,3}:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){1}(((:[0-9A-Fa-f]{1,4}){1,6})|((:[0-9A-Fa-f]{1,4}){0,4}:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:))|(:(((:[0-9A-Fa-f]{1,4}){1,7})|((:[0-9A-Fa-f]{1,4}){0,5}:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:)))(%.+)?s*(\/([0-9]|[1-9][0-9]|1[0-1][0-9]|12[0-8]))?$')) {
                    # Create Network Scan Object
                    $NetWorkScan = Invoke-PSnmap -ComputerName $subnet -Port $ports -Dns -NoSummary -AddService
                    # Filter devices that don't ping as no results will be found.
                    $scan = $NetworkScan | Where-Object { $_.Ping -eq $true }
                    Write-Verbose "##########################################"
                    Write-Verbose "Network scan for Subnet $Subnet completed."
                    Write-Verbose "DHCP Server: $($DHCPServer)"
                    Write-Verbose "Gateway: $($network.IPv4DefaultGateway.nexthop)"
                    Write-Verbose "##########################################"
                    $scan | ForEach-Object {
                        $org = ""
                        $macid = ((arp -a $_.ComputerName | Select-String '([0-9a-f]{2}-){5}[0-9a-f]{2}').Matches.Value).Replace("-", ":")
                        $macpop = $macid.replace(":", "")
                        $macsubstr = $macpop.Substring(0, 6)
                        $org = ($ouiobject | Where-Object { $_.assignment -eq $macsubstr })."Organization Name"
                        Add-Member -InputObject $_ -MemberType NoteProperty -Name MacID -Value $macid
                        if ($org) {
                            Add-Member -InputObject $_ -MemberType NoteProperty -Name ManufacturerName -Value $org
                        }
                        else {
                            Add-Member -InputObject $_ -MemberType NoteProperty -Name ManufacturerName -Value "Not Found"
                        }
                    }
                    # Normalize Subnet text for filename.
                    $subnetText = $(($subnet.Replace("/", ".CIDR.")))
                    # If report switch is true.
                    if ($report) {
                        $scan | Export-Csv "C:\temp\$((Get-Date).ToString('yyyy-MM-dd_hh.mm.ss')).$($env:USERDNSDOMAIN)_Subnet.$($subnetText)_DHCP.$($DHCPServer)_Gateway.$($network.IPv4DefaultGateway.nexthop).NetScan.csv" -NoTypeInformation
                    }
                    # Add scan to function output.
                    $results += $scan
                } # IF Subnet Match End
            } # End Foreach
        } # End If $LocalSubnets
        elseif ($Computers) {
            $Subnet = $Computers
            $results = Invoke-PSnmap -ComputerName $subnet -Port $ports -Dns -NoSummary -AddService | Where-Object { $_.Ping -eq $true }
            if ($Report) {
                $results | Export-Csv "C:\temp\$((Get-Date).ToString('yyyy-MM-dd_hh.mm.ss')).$($env:USERDNSDOMAIN)_HostScan.csv" -NoTypeInformation
            }
        }
    } # Process Close
    end {
        return $results
    }# End Close
}